ssl/quic/quic_rcidm.c

```diff
@@ -0,0 +1,679 @@
+ *       the RCID is not in the priority queue;
+ *       the RCID is not in the retiring_list.
+ *
+ *   RETIRING
+ *     Invariants:
+ *       rcid->state == RCID_STATE_RETIRING;
+ *       rcid->pq_idx == SIZE_MAX (debug assert only);
+ *       the RCID is not the current RCID, rcidm->cur_rcid != rcid;
+ *       the RCID is not in the priority queue;
+ *       the RCID is in the retiring_list.
+ *
+ *   Invariant: At most one RCID object is in the CURRENT state at any one time.
+ *
+ *      (If no RCID object is in the CURRENT state, this means either
+ *       an unnumbered RCID is being used as the preferred RCID
+ *       or we currently have no preferred RCID.)
+ *
+ *   All of the above states can be considered substates of the 'ACTIVE' state
+ *   for an RCID as specified in RFC 9000. A CID only ceases to be active
+ *   when we send a RETIRE_CONN_ID frame, which is the responsibility of the
+ *   user of the RCIDM and happens after the above state machine is terminated.
+ */
+enum {
+    RCID_STATE_PENDING,
+    RCID_STATE_CUR,
+    RCID_STATE_RETIRING
+};
+
+enum {
+    RCID_TYPE_INITIAL,      /* CID is from an peer INITIAL packet     (seq 0) */
+    RCID_TYPE_PREF_ADDR,    /* CID is from a preferred_address TPARAM (seq 1) */
+    RCID_TYPE_NCID          /* CID is from a NCID frame */
+    /*
+     * INITIAL_ODCID and RETRY_ODCID also conceptually exist but are tracked
+     * separately.
+     */
+};
+
+typedef struct rcid_st {
+    OSSL_LIST_MEMBER(retiring, struct rcid_st); /* valid iff retire == 1 */
+
+    QUIC_CONN_ID    cid;        /* The actual CID string for this RCID */
+    uint64_t        seq_num;
+    size_t          pq_idx;     /* Index of entry into priority queue */
+    unsigned int    state  : 2; /* RCID_STATE_* */
+    unsigned int    type   : 2; /* RCID_TYPE_* */
+} RCID;
+
+DEFINE_PRIORITY_QUEUE_OF(RCID);
+DEFINE_LIST_OF(retiring, RCID);
+
+/*
+ * RCID Manager
+ * ============
+ *
+ * The following "business logic" invariants also apply to the RCIDM
+ * as a whole:
+ *
+ *   Invariant: An RCID of INITIAL   type has a sequence number of 0.
+ *   Invariant: An RCID of PREF_ADDR type has a sequence number of 1.
+ *
+ *   Invariant: There is never more than one Initial ODCID
+ *              added throughout the lifetime of an RCIDM.
+ *   Invariant: There is never more than one Retry ODCID
+ *              added throughout the lifetime of an RCIDM.
+ *   Invariant: There is never more than one INITIAL RCID created
+ *              throughout the lifetime of an RCIDM.
+ *   Invariant: There is never more than one PREF_ADDR RCID created
+ *              throughout the lifetime of an RCIDM.
+ *   Invariant: No INITIAL or PREF_ADDR RCID may be added after
+ *              the handshake is completed.
+ *
+ */
+struct quic_rcidm_st {
+    /*
+     * The current RCID we prefer to use (value undefined if
+     * !have_preferred_rcid).
+     */
+    QUIC_CONN_ID                preferred_rcid;
```

Comment:
The distinction between the preferred_rcid and the cur_rcid is quite subtle. It would be helpful to have a comment somewhere explaining it.